File: src/RqCalc.WebApp/_Scripts/Models/Dialogs/BuffsWindowModel.js
function BuffsModel(buffs, startupBuffs) {

    var thisModel = this;

    thisModel.items = buffs.map(function (buff) {

        return {

            buff: buff,

            stacks: Enumerable.range(1, buff.maxStackCount).map(function (index) {

                return {

                    text: buff.bonuses.map(function (buffBonus) {

                        return TextTemplate.evaluate(buffBonus.template, [buffBonus.perStackValue * index]);
                    }).join("\r\n"),

                    index: index
                }
            }),

            value: ko.observable(startupBuffs.getValueByKeyOrDefault(buff))
        };
    });

    thisModel.clear = function() {
        
        thisModel.items.foreach(function (pair) {

            pair.value(null);
        });
    };
}


function BuffsWindowModel(context, character) {
    
    var thisModel = this;

    var startupBuffs = character.buffs.value();
    
    {
        var $class = character.class();
        var currentLevel = character.level();
        var currentTalents = character.talents.value();

        var classBuffs = $class.buffs.filter(function(buff) {
            return MainHelper.Buff.isAllowed(buff, currentLevel, currentTalents);
        });

        thisModel.classBuffs = new BuffsModel(classBuffs, startupBuffs);
    }
    
    {
        thisModel.sharedPositiveBuffs = new BuffsModel(context.sharedBuffs.filter(function (sharedBuff) {
            return !sharedBuff.isNegate;
        }), startupBuffs);
    }

    {
        thisModel.cardBuffs = new BuffsModel(MainHelper.CharacterSource.getCardBuffs(character), startupBuffs);
    }


    {
        thisModel.stampBuffs = new BuffsModel(MainHelper.CharacterSource.getStampBuffs(character), startupBuffs);
    }

    {
        thisModel.sharedNegativeBuffs = new BuffsModel(context.sharedBuffs.filter(function (sharedBuff) {
            return sharedBuff.isNegate;
        }), startupBuffs);
    }

    var models = [thisModel.classBuffs, thisModel.sharedPositiveBuffs, thisModel.cardBuffs, thisModel.stampBuffs, thisModel.sharedNegativeBuffs];
    
    thisModel.clear = function () {

        models.foreach(function(model) {
            model.clear();
        });
    };

    thisModel.getResult = function () {
        
        return models.selectMany(function (model) {
            return model.items;
        }).map(function (pair) {
            return {
                key: pair.buff,
                value: pair.value()
            }
        }).filter(function (pair) {
            return pair.value != null;
        });
    }
}

File: src/RqCalc.Wpf/Models/Window/Dialog/BuffsWindowModel.cs
using Framework.Core;
using Framework.DataBase;
using Framework.HierarchicalExpand;
using Framework.Reactive;

using RqCalc.Domain;
using RqCalc.Domain._Base;
using RqCalc.Domain._Extensions;
using RqCalc.Model;
using RqCalc.Model._Extensions;
using RqCalc.Wpf.Models.Window.Dialog._Base;

namespace RqCalc.Wpf.Models.Window.Dialog;

public class BuffsWindowModel : NotifyModelBase, IClearModel
{
    public BuffsWindowModel(IDataSource<IPersistentDomainObjectBase> dataSource, IVersion lastVersion, ICharacterSource character)
    {
        this.ClassBuffList = character.Class.GetAllParents()
                                      .Reverse()
                                      .SelectMany(c => c.Buffs)
                                      .WhereVersion(lastVersion)
                                      .Where(buff => buff.IsAllowed(character.Level, character.Talents))
                                      .ToObservableCollection(buff => new BuffModel(buff));

        var sharedBuffList = dataSource
            .GetFullList<IBuff>()
            .WhereVersion(lastVersion)
            .Where(buff => buff.IsShared())
            .ToObservableCollection(buff => new BuffModel(buff));

        this.SharedPositiveBuffList = sharedBuffList.Where(buff => !buff.SelectedObject.IsNegate).ToObservableCollection();

        this.SharedNegativeBuffList = sharedBuffList.Where(buff => buff.SelectedObject.IsNegate).ToObservableCollection();

        this.CardBuffList = character.GetCardBuffs().ToObservableCollection(buff => new BuffModel(buff));

        this.StampBuffList = character.GetStampBuffs().ToObservableCollection(buff => new BuffModel(buff));

        this.Buffs = character.Buffs;
    }




    public IReadOnlyDictionary<IBuff, int> Buffs
    {
        get => this.ClassBuffs.Concat(this.SharedBuffs).Concat(this.CardBuffs).Concat(this.StampBuffs);
        private set
        {
            this.ClassBuffs = value.Where(pair => pair.Key.Class != null).ToDictionary();
            this.SharedBuffs = value.Where(pair => pair.Key.IsShared()).ToDictionary();
            this.CardBuffs = value.Where(pair => pair.Key.Card != null).ToDictionary();
            this.StampBuffs = value.Where(pair => pair.Key.Stamp != null).ToDictionary();
        }
    }




    public IReadOnlyDictionary<IBuff, int> ClassBuffs
    {
        get => this.ClassBuffList.Where(model => model.Value != 0).ToDictionary(pair => pair.SelectedObject, pair => pair.Value);
        private set
        {
            foreach (var model in this.ClassBuffList)
            {
                model.Value = value.GetMaybeValue(model.SelectedObject).GetValueOrDefault();
            }
        }
    }

    public ObservableCollection<BuffModel> ClassBuffList
    {
        get => this.GetValue(v => v.ClassBuffList);
        private set => this.SetValue(v => v.ClassBuffList, value);
    }



    public IReadOnlyDictionary<IBuff, int> SharedBuffs
    {
        get => this.SharedPositiveBuffList.Concat(this.SharedNegativeBuffList).Where(model => model.Value != 0).ToDictionary(pair => pair.SelectedObject, pair => pair.Value);
        private set
        {
            foreach (var model in this.SharedPositiveBuffList.Concat(this.SharedNegativeBuffList))
            {
                model.Value = value.GetMaybeValue(model.SelectedObject).GetValueOrDefault();
            }
        }
    }

    public ObservableCollection<BuffModel> SharedPositiveBuffList
    {
        get => this.GetValue(v => v.SharedPositiveBuffList);
        private set => this.SetValue(v => v.SharedPositiveBuffList, value);
    }

    public ObservableCollection<BuffModel> SharedNegativeBuffList
    {
        get => this.GetValue(v => v.SharedNegativeBuffList);
        private set => this.SetValue(v => v.SharedNegativeBuffList, value);
    }


    public IReadOnlyDictionary<IBuff, int> CardBuffs
    {
        get => this.CardBuffList.Where(model => model.Value != 0).ToDictionary(pair => pair.SelectedObject, pair => pair.Value);
        private set
        {
            foreach (var model in this.CardBuffList)
            {
                model.Value = value.GetMaybeValue(model.SelectedObject).GetValueOrDefault();
            }
        }
    }

    public ObservableCollection<BuffModel> CardBuffList
    {
        get => this.GetValue(v => v.CardBuffList);
        private set => this.SetValue(v => v.CardBuffList, value);
    }

    public IReadOnlyDictionary<IBuff, int> StampBuffs
    {
        get => this.StampBuffList.Where(model => model.Value != 0).ToDictionary(pair => pair.SelectedObject, pair => pair.Value);
        private set
        {
            foreach (var model in this.StampBuffList)
            {
                model.Value = value.GetMaybeValue(model.SelectedObject).GetValueOrDefault();
            }
        }
    }

    public ObservableCollection<BuffModel> StampBuffList
    {
        get => this.GetValue(v => v.StampBuffList);
        private set => this.SetValue(v => v.StampBuffList, value);
    }


    public bool CloseDialog => false;

    public virtual void Clear() => this.Buffs = new Dictionary<IBuff, int>();
}